pkg/jobstore/boltdb/context.go

```diff
@@ -62,28 +69,37 @@ func txFromContext(ctx context.Context) (*bolt.Tx, bool) {
 // Commit commits the transaction and cancels the context.
 // Commit will return an error if the transaction is already committed or rolled back.
 func (b *txContext) Commit() error {
-	defer b.cancelFunc()
 	b.mu.Lock()
 	defer b.mu.Unlock()
+	defer b.close()
 	return b.tx.Commit()
 }
 
 // Rollback rolls back the transaction and cancels the context.
 // Rollback is a no-op if the transaction is already committed or rolled back.
 func (b *txContext) Rollback() error {
-	defer b.cancelFunc()
 	return b.doRollback()
 }
 
 // doRollback is a helper function to rollback the transaction without cancelling the context.
 func (b *txContext) doRollback() error {
 	b.mu.Lock()
 	defer b.mu.Unlock()
+	defer b.close()
 	if err := b.tx.Rollback(); err != nil && !errors.Is(err, bolt.ErrTxClosed) {
 		return err
 	}
 	return nil
 }
 
+// close closes the transactional context.
+// already called with the mutex held.
+func (b *txContext) close() {
+	if !b.closed {
+		close(b.closeCh)
+		b.closed = true
+	}
+}
+
```

Comment:
since this logic should not be called without locking my preference would be to inline the logic of this function into the various places it's used and drop the function.